lib/cn/cn.c

```diff
@@ -1329,13 +1298,25 @@ cn_work_wrapper(struct work_struct *context)
 void
 cn_work_submit(struct cn *cn, cn_work_fn *handler, struct cn_work *work)
 {
-    cn_ref_get(cn);
+    struct workqueue_struct *wq;
 
     work->cnw_cnref = cn;
     work->cnw_handler = handler;
 
     INIT_WORK(&work->cnw_work, cn_work_wrapper);
-    queue_work(cn->cn_maint_wq, &work->cnw_work);
+
+    if (!cn) {
+        handler(work);
+        return;
+    }
+
+    cn_ref_get(cn);
+
+    wq = cn_get_maint_wq(cn);
+    if (wq)
+        queue_work(wq, &work->cnw_work);
+    else
+        cn_work_wrapper(&work->cnw_work);
```

Comment:
I converted cn_tree_destroy() to use cn_work_submit() rather than create its own workqueue (to uniqify this work-submit mechanism).  But some calls to cn_tree_destroy() are for a tree that hasn't been tied to a cn yet so needed to handle the nil cn case.